File: basic/singly_linked_list.js
const SinglyLinkedListNode = class {
  constructor(nodeData) {
    this.data = nodeData;
    this.next = null;
  }
};

const SinglyLinkedList = class {
  construction() {
    this.head = null;
    this.tail = null;
  }

  insertNode(nodeData) {
    const node = new SinglyLinkedListNode (nodeData);

    if (this.head == null) {
      this.head = node;
    } else {
      this.tail.next = node;
    }
    this.tail = node
  }
};

function printSinglyLinkedList(node) {
  while (node != null) {
    console.log(String(node.data));
    node = node.next;
  }
}

module.exports = {
  SinglyLinkedList,
  SinglyLinkedListNode,
  printSinglyLinkedList
};

File: basic/singly_linked_list.py
class SinglyLinkedListNode:
  def __init__(self, data):
    self.data = data
    self.next = None
  
class SinglyLinkedList:
  def __init__(self):
    self.head = None
    self.tail = None
  
  def insertNode(self, data):
    node = SinglyLinkedListNode(data)
    if self.head == None:
      self.head = node
    else:
      self.tail.next = node

    self.tail = node
  def printList(self):
    node = self.head
    result = ''
    while(node):
      result += str(node.data) + ' '
      node = node.next
    print(result)
  def reverse(self):
    currNode = self.head
    nextNode, prevNode = None, None
    while(currNode):
      nextNode = currNode.next
      currNode.next = prevNode
      prevNode = currNode
      currNode = nextNode
    self.head = prevNode

arr = [1, 2, 32, 4]
llist = SinglyLinkedList()
for a in arr:
  llist.insertNode(a)
llist.reverse()
llist.printList()
